texar/modules/encoders/xlnet_encoder.py

```diff
@@ -0,0 +1,480 @@
+
+        `initializer`: dict, optional
+            Hyperparameters of the default initializer that initializes
+            variables created in this module.
+            See :func:`~texar.core.get_initializer` for details.
+
+        `name`: str
+            Name of the module.
+        """
+
+        return {
+            'pretrained_model_name': 'xlnet-base-cased',
+            'untie_r': True,
+            'num_layers': 12,
+            'mem_len': 0,
+            'reuse_len': 0,
+            # layer
+            'num_heads': 12,
+            'hidden_dim': 768,
+            'head_dim': 64,
+            'dropout': 0.1,
+            'attention_dropout': 0.1,
+            'use_segments': True,
+            # ffn
+            'ffn_inner_dim': 3072,
+            'activation': 'gelu',
+            # embedding
+            'vocab_size': 32000,
+            'max_seq_len': 512,
+            'initializer': None,
+            'name': "xlnet_encoder",
+            '@no_typecheck': ['pretrained_model_name'],
+        }
+
+    @property
+    def output_size(self):
+        return self._hparams.hidden_dim
+
+    @staticmethod
+    def _cache_mem(output: torch.Tensor,
+                   prev_mem: Optional[torch.Tensor],
+                   mem_len: int,
+                   reuse_len: int = 0) -> torch.Tensor:
+        r"""Cache hidden states into memory."""
+        assert mem_len > 0
+
+        if reuse_len is not None and reuse_len > 0:
+            output = output[:reuse_len]
+        if prev_mem is None:
+            new_mem = output[-mem_len:]
+        else:
+            new_mem = torch.cat([prev_mem, output], dim=0)[-mem_len:]
+        return new_mem.detach()
+
+    def _create_causal_attn_mask(self,
+                                 seq_len: int,
+                                 mem_len: int,
+                                 same_length: bool = False) -> torch.Tensor:
+        r"""Create causal attention mask of shape
+        `(seq_len, mem_len + seq_len)`.
+        """
+        assert self.r_w_bias is not None
+        device = self.r_w_bias.device
+        attn_mask = torch.ones(seq_len, seq_len, device=device)
+        mask_u = torch.triu(attn_mask, diagonal=1)
+        attn_mask_pad = torch.zeros(seq_len, mem_len, device=device)
+        ret = torch.cat([attn_mask_pad, mask_u], dim=1)
+        if same_length:
+            mask_l = torch.tril(attn_mask, diagonal=-1)
+            ret = torch.cat([ret[:, :seq_len] + mask_l, ret[:, seq_len:]], 1)
+        return ret
+
+    def forward(self,  # type: ignore
+                token_ids: torch.LongTensor,
+                *args,
+                **kwargs) \
+            -> Tuple[torch.Tensor, Optional[List[torch.Tensor]]]:
+        r"""A wrapper for :meth:`_forward`. This layer exists because
+        :class:`XLNetDecoder` compute embeddings in the decoder helper.
+        Please refer to :meth:`_forward` for the full list of arguments.
```

Comment:
Does this mean `XLNetDecoder` shares `_forward` w/ `XLNetEncoder`?

If so, this is kind of internal implementation details and perhaps should not be exposed to users. I'd suggest moving the docstring of `_forward` to `forward`. And `_forward` is not visible to users.

Reply:
Yes -- the reason to have a separate `_forward` method was because `Helper` returns the embeddings instead of tokens. I agree we should not expose internal methods and we should move the docstring.